Add fixers and more output

roguh opened the pull request:
This will add more logging, such as printing the command before running it or printing `[no output]`. Also adds the option to add multiple linters per file type.

I also added the ability to run `fixers`, which may modify the file. This can be used for formatters or linters with a `--fix` option, like `ruff --fix` and `orange` for Python, `cargo clippy --fix` for Rust, or `jq -c` for JSON.

roguh committed:
More output and isolate code in run_actions_on_file function

roguh committed:
Add fixers that could modify the file

roguh committed:
Add Rust linters/fixers

rnpnr commented:
Hi Hugo, thanks for the patch! Some comments:

- I made a change to pipe the file contents to the command so that you don't have to write the file before linting. Can you adjust your patch to maintain that behavior if possible?
- Please don't switch tabs to spaces. That way I can have my 8 char tabstop and you can have your 4 char tabstop and we are both happy.

More comments inline:

rnpnr commented on init.lua:
```diff
@@ -1,21 +1,51 @@
+        local _, ostr, estr = vis:pipe(file, {start = 0, finish = 0}, full_cmd)
+
+        if ostr == nil and estr == nil then
+          vis:message('[no output]')
```
Ideally this should go to the info bar and only when all of the commands have no output.

roguh replied:
I'll just remove that.

rnpnr commented on init.lua:
```diff
@@ -1,21 +1,51 @@
+fixers = {}
+fixers["python"] = {"black", "isort", "ruff --fix"}
+fixers["json"] = {"jq -c"}
+fixers["rust"] = {"cargo fmt", "cargo clippy --fix"}
```
Please keep these lists sorted.

rnpnr commented on init.lua:
```diff
@@ -1,21 +1,51 @@
+    end
+    if file.name == nil then
+        vis:info(action .. " cannot run on nil file")
+        return false
```
This shouldn't be necessary if contents from the buffer are piped to the command.

rnpnr commented:
I also forgot to add that I am going to assume that the commands you have added are somewhat "standard" since I don't really use any of those languages. When I have time I would like to make the linters configurable with these being the defaults.

rnpnr committed:
pipe to and from fixers

rnpnr commented:
Hi Hugo,
I cleaned up the main issues I had. This can be merged if I can get some confirmation that the linters and fixers you added accept file contents on `stdin` and write the results to `stdout`.

roguh commented:
Hey, sorry for dropping off, life got quite busy. I'll double check that the commands are standard and work well stdin and stdout for I/O.

Piping output into commands will handle the edge case of opening `stdin`, e.g. as an interactive editor within a sequence of piped commands. Mainly, it'll avoid problems with temporary files or writing to disk automatically. Good idea!

> Please don't switch tabs to spaces. That way I can have my 8 char tabstop and you can have your 4 char tabstop and we are both happy.

Got it!

roguh committed:
Add Rust and Python linters and fixers that take input from stdin

roguh committed:
Cleaner and quicker vis:message output, and run all linters even if one fails

- Less newlines
- No more `$ [cmd]...`. Use description that indicates file is piped to the given command.
  since we are now piping to the commands, not running shell commands.
- Show output as soon as it's ready instead of waiting on all linters/fixers.
- Fix bug if vis.win.syntax is undefined
- Print which file is being worked on
- Print header before most of our plugin's output: such as "--- linters: " ...
- Print message when done
- Use fixer status code to see if the fixer failed and exit early
  Do not use stderr, since some fixers print to stderr even on success

rnpnr commented:
These changes look good to me. I'm fine with merging as is now.

I'm going to make the fixers and linters configurable soon. When I do so I think I only want to keep at most 2 default fixers/linters. Do you have a preference for `python`? Maybe `black` and `isort`?

It will be really easy to add more to the defaults in `visrc.lua` so don't worry about losing the others.

roguh commented:
For Python, I suggest `black`, `isort`, and `pylint`.

Can you leave the other ones in there somewhere? I'd want users to have these commands that are compatible with stdin/stdout but only a few should be enabled by default. The `mypy <(cat)` command is especially esoteric. Perhaps a table mapping a language + linter/fixer and its stdout/stdin command?

roguh commented:
I wrote up a little something so other people can find this `mypy <(cat)` solution: https://felina.cloud/how-to-send-stdin-to-mypy
